onnx/defs/nn/defs.cc

```diff
@@ -2595,4 +2595,151 @@ ONNX_OPERATOR_SET_SCHEMA(
+            1,
+            OpSchema::Differentiable)
+        .Input(
+            2,
+            "bias",
+            "Bias tensor of shape `(num_groups)`.",
+            "T",
+            OpSchema::Single,
+            true,
+            1,
+            OpSchema::Differentiable)
+        .Output(
+            0,
+            "Y",
+            "The output tensor of the same shape as `X`.",
+            "T",
+            OpSchema::Single,
+            true,
+            1,
+            OpSchema::Differentiable)
+        .TypeConstraint(
+            "T",
+            {"tensor(float16)", "tensor(float)", "tensor(double)", "tensor(bfloat16)"},
+            "Constrain input and output types to float tensors.")
+        .SetContextDependentFunctionBodyBuilder(
+            [](const FunctionBodyBuildContext& ctx, const OpSchema& schema, FunctionProto& functionProto) {
+              // GroupNormalization <epsilon, num_groups> (X, scale, bias) => (Y)
+              auto* tp = ctx.getInputType(0);
+              if ((tp == nullptr) || (!tp->has_tensor_type()))
+                return false;
+              int64_t T = tp->tensor_type().elem_type();
+
+              auto* epsilon_attr = ctx.getAttribute("epsilon");
+              float epsilon = (epsilon_attr != nullptr) ? epsilon_attr->f() : 1e-5f;
+              auto* num_groups_attr = ctx.getAttribute("num_groups");
+              if (num_groups_attr == nullptr)
+                return false;
+              int64_t num_groups = num_groups_attr->i();
+
+              auto& input_shape = ctx.getInputType(0)->tensor_type().shape();
+              int64_t input_ndim = input_shape.dim_size();
+              std::vector<int64_t> reduce_axes;
+              for (int64_t i = 2; i < input_ndim + 1; i++) {
+                reduce_axes.push_back(i);
+              }
+
+              auto mktensor = [](int64_t val) -> ONNX_NAMESPACE::TensorProto {
+                auto tp = ONNX_NAMESPACE::ToTensor(std::vector<int64_t>{val});
+                tp.add_dims(1);
+                return tp;
+              };
+              FunctionBuilder builder(functionProto);
+              builder.Const1D("FloatEpsilon", epsilon)
+                  .Add("Epsilon = Cast (FloatEpsilon)", "to", T)
+                  .Add("XShape = Shape (X)") // shape of input tensor: 1D tensor
+                  .Add("Rank = Size (XShape)") // rank of input tensor: scalar
+                  .Const1D("One1D", int64_t(1))
+                  .Const1D("Two1D", int64_t(2))
+                  .Add("C = Slice (XShape, One1D, Two1D)") // number of channels
+                  .Const1D("NumGroups", num_groups)
+                  .Add("GroupSize = Div (C, NumGroups)")
+                  .Const1D("Zero1D", int64_t(0))
+                  .Const1D("IntMax1D", std::numeric_limits<int64_t>::max())
+                  .Add("N = Slice (XShape, Zero1D, One1D)") // batch size
+                  .Add("HW = Slice (XShape, Two1D, IntMax1D)") // data instance shape
```

Comment:
And `HW = Shape <start=2>(X)` should work here. Also renaming `HW` to `Rest` will probably more accurately describe the non-image cases.